docs: query engine user guide (#250)
limebit/medmodels: docs/user_guide/02b_query_engine.md

- [`has_attribute()`](medmodels.medrecord.querying.NodeOperand.has_attribute){target="_blank"} : Query nodes that have that attribute.
- [`attribute()`](medmodels.medrecord.querying.NodeOperand.attribute){target="_blank"} : Returns a [`MultipleValuesOperand`](medmodels.medrecord.querying.MultipleValuesOperand){target="_blank"} to query on the values of the nodes for that attribute.
- [`greater_than()`](medmodels.medrecord.querying.MultipleValuesOperand.greater_than){target="_blank"}` : Query values that are greater than that value.
- [`select_nodes()`](medmodels.medrecord.medrecord.MedRecord.select_nodes){target="_blank"} : Select nodes that match that query.

:::

:::{note}
The [`has_attribute()`](medmodels.medrecord.querying.NodeOperand.has_attribute){target="_blank"} method is not needed in this example, since the [`attribute()`](medmodels.medrecord.querying.NodeOperand.attribute){target="_blank"} one already checks whether the nodes have the attribute. It is there merely for educational purposes. This will happen in different examples in this user guide to ensure the maximum amount of methods are portrayed.
:::

> In case, for instance, that you do not know whether there are different ways to assign the `gender` attribute across the [`MedRecord`](medmodels.medrecord.medrecord.MedRecord){target="_blank"} (with leading/trailing whitespaces or formatted in lower/uppercase), you can also modify the attributes of a node/edge. You can also perform mathematical calculations like [`mean()`](medmodels.medrecord.querying.MultipleValuesOperand.mean){target="_blank"}, [`median()`](medmodels.medrecord.querying.MultipleValuesOperand.median){target="_blank"} or [`min()`](medmodels.medrecord.querying.MultipleValuesOperand.min){target="_blank"} and assign them to a variable. Also, you can keep manipulating the operand, like in the following example, where we are subtracting _5_ years from the `mean_age` to query on that value.

Reviewer: > [...] you can also modify the attributes of a node/edge. You can also perform mathematical [...]

nitpick: you can also is repeated here

- [`less_than()`](medmodels.medrecord.querying.MultipleValuesOperand.less_than){target="_blank"} : Query values that are less than that value.
- [`select_nodes()`](medmodels.medrecord.medrecord.MedRecord.select_nodes){target="_blank"} : Select nodes that match that query.

:::

:::{note}
Query methods used for changing the operands cannot be concatenated or assigned to variables, since their `Return` is null. That is, the following code snippet will set `gender_lowercase` as `None`, and thus the query will not be able to find any nodes.

```python
# Wrong implementation
gender_lowercase = node.attribute("gender").lowercase()
gender.equal_to("m")

# Wrong implementation
gender = node.attribute("gender")
gender.lowercase().trim()
gender.equal_to("m")

# Correct implementation
gender = node.attribute("gender")
gender.lowercase()
gender.trim()
gender.equal_to("m")
```

Nor do the ones that compare operands to other operands, since their `Return` value is also null.

```python
# Wrong implementation
gender = node.attribute("gender")
gender.equal_to("M").not_equal_to("F")

# Correct implementation
gender = node.attribute("gender")
gender.equal_to("M")
gender.not_equal_to("F")
```

> :::

Reviewer: very important part, noice!

---

the return value is `None` and not `null`. I also don't think I'd put return (or Return) in the `code comment mode`

---

did you mean
```python
# Wrong implementation
gender_lowercase = node.attribute("gender").lowercase()
gender_lowercase.equal_to("m")
```
?

---

I also wouldn't say that

> [...] the query will not be able to find any nodes.

as these queries would result in a `AttributeError` to be thrown, e.g.
```python
AttributeError("'NoneType' object has no attribute 'equal_to'")
```

> As you can see, the query engine is highly useful for finding nodes that fulfill different criteria in a highly optimized way, using descriptive programming as a precise tool. We can use previously defined queries too, and also use the [`neighbors()`](medmodels.medrecord.querying.NodeOperand.neighbors){target="_blank"} method to query also through the nodes that are neighbors to those nodes.

Reviewer: > As you can see, the query engine is highly useful for finding nodes that fulfill different criteria in a highly optimized way, using descriptive programming as a precise tool.

This sounds very chatgpt-ey, especially the last part.

---

> We can use previously defined queries too, and also use the neighbors() method to query also through the nodes that are neighbors to those nodes.

This sentence feels a bit weird as you first explain that queries can be re-used (which should be highlighted much more prominently imo) and then in the next sub-clause introduce the `neighbors` functionality which is a completely different topic. Also there is one `also` too much.

For instance, in the following example we are selecting the nodes that have the following characteristics

- Are in group `patient`.
- Their node index contains the string _"pat"_
- Their attribute `age` is greater than 30, and their attribute `gender` is equal to _"M"_.

> - They are  connected to  nodes which attribute `description` contains the word _"fentanyl"_  in either upper or lowercase.

Reviewer: ```suggestion
- They are connected to nodes whose attribute `description` contains the word _"fentanyl"_ in either uppercase or lowercase.
```

```{exec-literalinclude} scripts/02b_query_engine.py
---
language: python
setup-lines: 1-4, 15-24
lines: 46-55
---
```

:::{dropdown} Methods used in the snippet

- [`neighbors()`](medmodels.medrecord.querying.NodeOperand.neighbors){target="_blank"} : Returns a [`NodeOperand()`](medmodels.medrecord.querying.NodeOperand){target="_blank"} to query the neighbors of those nodes.
- [`attribute()`](medmodels.medrecord.querying.NodeOperand.attribute){target="_blank"} : Returns a [`MultipleValuesOperand()`](medmodels.medrecord.querying.MultipleValuesOperand){target="_blank"} to query on the values of the nodes for that attribute.
- [`lowercase()`](medmodels.medrecord.querying.MultipleValuesOperand.lowercase){target="_blank"} : Converts the values that are strings to lowercase.
- [`contains()`](medmodels.medrecord.querying.NodeIndexOperand.contains){target="_blank"} : Query node indices containing that argument.
- [`select_nodes()`](medmodels.medrecord.medrecord.MedRecord.select_nodes){target="_blank"} : Select nodes that match that query.

:::

## Edge Queries

> Edge operands provide a way to define and filter relationships between nodes in a [`MedRecord`](medmodels.medrecord.medrecord.MedRecord){target="_blank"}. By using edge operands, you can create queries that target specific connections based on attributes like time, duration, or custom properties. This section explores how to leverage edge operands to gain insights from the relationships in your data. For that, the [`EdgeOperand`](medmodels.medrecord.querying.EdgeOperand){target="_blank"} querying class is used.

Reviewer: > Edge operands provide a way to define and filter relationships between nodes in a [...]

should be edges not nodes.

---

> By using edge operands, you can create queries that target specific connections based on attributes like time, duration, or custom properties.

Sounds very limiting as edges has basically the same functionalities in this regard (and also the others) as the `NodeOperand`

---

> This section explores how to ✨leverage✨ edge operands to ✨gain insights✨ from [...]

also kinda chatgpt-ey

# Query Engine

## What is the Query Engine?

The **MedRecord Query Engine** enables users to find node and edges' indices stored in the graph structure efficiently. Thanks to an intuitive interface, complex queries can be performed, allowing you to filter of nodes and edges by their properties and relationships. This section introduces the basic concepts of querying MedRecords and explores advanced techniques for working with complex datasets.

## Example dataset

An example dataset can be used using the [`MedRecord`](medmodels.medrecord.medrecord.MedRecord){target="_blank"}'s method [`from_example_dataset`](medmodels.medrecord.medrecord.MedRecord.from_example_dataset){target="_blank"}.

```{literalinclude} scripts/02b_show_dataset.py
---
language: python
lines: 8
---
```

> This example dataset includes a set of patients, drugs, diagnoses and procedures. For this section, we will only use the patients, drugs and the edges that connect these two groups.

Reviewer 2: `show()` is a local function defined in the script `scripts/02b_show_dataset.py` right? I think for the user it looks like `show()` is something they could use anywhere and that is misleading I guess. I suggest to either: 
- make show a general function
- Use `print()` in the user guide 
- Explain how `show()` is defined

```{exec-literalinclude} scripts/02b_show_dataset.py
---
language: python
setup-lines: 1-42
lines: 44
---
```

```{exec-literalinclude} scripts/02b_show_dataset.py
---
language: python
setup-lines: 1-42
lines: 45
---
```

```{exec-literalinclude} scripts/02b_show_dataset.py
---
language: python
setup-lines: 1-42
lines: 46
---
```

## Node Queries

> The [`NodeOperand`](medmodels.medrecord.querying.NodeOperand){target="_blank"} querying class allow you to define specific criteria for selecting nodes within a [`MedRecord`](medmodels.medrecord.medrecord.MedRecord){target="_blank"}. These operands enable flexible and complex queries by combining multiple conditions, such as group membership, attributes' selection and querying, attribute values, and relationships to other nodes or edges. This section introduces the basic usage of node operands to create a powerful foundation for your data queries.

Reviewer 2: The ' in attributes' is not necessary, right?

> The [`NodeOperand`](medmodels.medrecord.querying.NodeOperand){target="_blank"} querying class allow you to define specific criteria for selecting nodes within a [`MedRecord`](medmodels.medrecord.medrecord.MedRecord){target="_blank"}. These operands enable flexible and complex queries by combining multiple conditions, such as group membership, attributes' selection and querying, attribute values, and relationships to other nodes or edges. This section introduces the basic usage of node operands to create a powerful foundation for your data queries.

Reviewer 2: I am not yet fully convinced about the naming of the queries. They are named: `query_node_basic `, `query_node_intermmediate`, `query_node_advanced`. Why that? From a users perspective, wouldn't it be more clean to read if the queries are named more like whats in it? So the query with the content `node.in_group("patient")` could be named for something that is more expressive to that content and the user is made aware that naming can be set to his or her choice.

```{exec-literalinclude} scripts/02b_query_engine.py
---
language: python
setup-lines: 1-4
lines: 7-12
---
```

> :::{dropdown} Methods used in the snippet

Reviewer 2: Thats super cool!

```{exec-literalinclude} scripts/02b_query_engine.py
---
language: python
setup-lines: 1-12
lines: 27-36
---
```

:::{dropdown} Methods used in the snippet

- [`index()`](medmodels.medrecord.querying.NodeOperand.index){target="_blank"}: Returns a [`NodeIndexOperand`](medmodels.medrecord.querying.NodeIndexOperand){target="_blank"}` to query on the indices.
- [`contains()`](medmodels.medrecord.querying.NodeIndexOperand.contains){target="_blank"} : Query node indices containing that argument.
- [`has_attribute()`](medmodels.medrecord.querying.NodeOperand.has_attribute){target="_blank"} : Query nodes that have that attribute.
- [`attribute()`](medmodels.medrecord.querying.NodeOperand.attribute){target="_blank"} : Returns a [`MultipleValuesOperand`](medmodels.medrecord.querying.MultipleValuesOperand){target="_blank"} to query on the values of the nodes for that attribute.
- [`greater_than()`](medmodels.medrecord.querying.MultipleValuesOperand.greater_than){target="_blank"}` : Query values that are greater than that value.
- [`select_nodes()`](medmodels.medrecord.medrecord.MedRecord.select_nodes){target="_blank"} : Select nodes that match that query.

:::

### Advanced Query Operations

> In case, for instance, that you do not know whether there are different ways to assign the `gender` attribute across the [`MedRecord`](medmodels.medrecord.medrecord.MedRecord){target="_blank"} (with leading/trailing whitespaces or formatted in lower/uppercase), you can modify the value of the attributes of a node/edge inside the query. It is important to note that modifying these values does not change the actual value of the attributes within the [`MedRecord`](medmodels.medrecord.medrecord.MedRecord){target="_blank"}: it just changes the value of those variables in the query.

Reviewer 2: The sentence 

> It is important to note that modifying these values does not change the actual value of the attributes within the [`MedRecord`](medmodels.medrecord.medrecord.MedRecord){target="_blank"}: it just changes the value of those variables in the query.

I quite important I guess, maybe we can highlight it a bit more?

- [`select_nodes()`](medmodels.medrecord.medrecord.MedRecord.select_nodes){target="_blank"} : Select nodes that match that query.

:::

This includes also _"pat_3"_, that was not included in the previous section because none of its edges was included in the `query_edge_either()`, but it can be found in the `query_edge_or()` now.

```{exec-literalinclude} scripts/02b_query_engine.py
---
language: python
setup-lines: 1-4, 109-129
lines: 133-139
---
```

:::{dropdown} Methods used in the snippet

- [`in_group()`](medmodels.medrecord.querying.EdgeOperand.in_group){target="_blank"} : Query nodes that belong to that group.: Query edges that belong to that group.
- [`exclude()`](medmodels.medrecord.querying.NodeOperand.exclude){target="_blank"} : Exclude the nodes that belong to the given query.
- [`select_nodes()`](medmodels.medrecord.medrecord.MedRecord.select_nodes){target="_blank"} : Select nodes that match that query.

:::

So this gives us all the patient nodes that were not selected with the previous query (logical **NOT** applied).

## Clones

> Since the statements in the query engine are additive, we cannot go back to a previous state of the query unless we want to rewrite the whole query again for an intermemediate step. For that reason, [`clone()`](medmodels.medrecord.querying.SingleValueOperand.clone){target="_blank"} method was devised. Clones of all types of operands can be made.

Reviewer 2: intermemediate -> intermediate step

> Since the statements in the query engine are additive, we cannot go back to a previous state of the query unless we want to rewrite the whole query again for an intermemediate step. For that reason, [`clone()`](medmodels.medrecord.querying.SingleValueOperand.clone){target="_blank"} method was devised. Clones of all types of operands can be made.

Reviewer 2: I think the script in this part requires an additional code comment to explain what is happening there. This part of the chapter is a bit short and not easy enough to understand. So maybe change the script a bit and show that you cloned something and then went on also with the original value / intermediate step

- [`greater_than()`](medmodels.medrecord.querying.MultipleValuesOperand.greater_than){target="_blank"} : Query values that are greater than that value.
- [`less_than()`](medmodels.medrecord.querying.MultipleValuesOperand.less_than){target="_blank"} : Query values that are less than that value.
- [`select_nodes()`](medmodels.medrecord.medrecord.MedRecord.select_nodes){target="_blank"} : Select nodes that match that query.

:::

## Queries as Function Arguments

In all previous snippets, we have used queries with the method [`select_nodes()`](medmodels.medrecord.medrecord.MedRecord.select_nodes){target="_blank"} for representation purposes of its capacities. However, queries can also be used as function arguments to other methods or indexers from the [`MedRecord`](medmodels.medrecord.medrecord.MedRecord){target="_blank"}:

- Using [`node[]`](medmodels.medrecord.medrecord.MedRecord.node){target="_blank"}, an indexer that retrieves the attributes for the given node indices:

```{exec-literalinclude} scripts/02b_query_engine.py
---
language: python
setup-lines: 1-153
lines: 158
---
```

> - Using [`groups_of_node()`](medmodels.medrecord.medrecord.MedRecord.groups_of_node){target="_blank"}, a method that retrieves the groups to which a specific node index belongs to.

Reviewer 2: I think at the end of the chapter, so roundabout here, the explanations are a bit speeded up. Maybe have one or two more lines in this part to explain what is happening here

Author: Correct me if I am wrong btu I do not think it should be extended more, since these methods are not part of the query engine per se. They are different functions of the MedRecord that could take the queries as arguments.

Reviewer 2: Agreed, and the method is linked, therefore that's okay

```{literalinclude} scripts/02b_show_dataset.py
---
language: python
lines: 7
---
```

This example dataset includes a set of patients, drugs, diagnoses and procedures. For this section, we will only use the patients, drugs and the edges that connect these two groups.

```{exec-literalinclude} scripts/02b_show_dataset.py
---
language: python
setup-lines: 1-36
lines: 38
---
```

```{exec-literalinclude} scripts/02b_show_dataset.py
---
language: python
setup-lines: 1-36
lines: 39
---
```

```{exec-literalinclude} scripts/02b_show_dataset.py
---
language: python
setup-lines: 1-36
lines: 40
---
```

## Node Queries

The [`NodeOperand`](medmodels.medrecord.querying.NodeOperand){target="_blank"} querying class allow you to define specific criteria for selecting nodes within a [`MedRecord`](medmodels.medrecord.medrecord.MedRecord){target="_blank"}. These operands enable flexible and complex queries by combining multiple conditions, such as group membership, attributes' selection and querying, attribute values, and relationships to other nodes or edges. This section introduces the basic usage of node operands to create a powerful foundation for your data queries.

> ```{exec-literalinclude} scripts/02b_query_engine.py

Author: It appears at the end of the page, in the full code section. Should I still include it anyways in each snippet?

> ```{exec-literalinclude} scripts/02b_query_engine.py

Reviewer 2: Its okay for now. I will resctructure things anyway a bit as discussed, so we can postpone this topic.
